torch/nn/modules/relation.py

```diff
@@ -0,0 +1,47 @@
+import torch
+from .module import Module
+from torch.autograd import Variable
+
+
+class Relation(Module):
+  def __init__(self, f, g):
+    r"""Applies a relation network to the incoming data: :math:`y = f\left(\sum_{i,j} g(x_i, x_j)\right)`
+
+    Args:
+        f: function/network applied to the sum of the pairs of objects applied to g
+        g: function/network applied to the pairs of objects x
+
+    Shape:
+        - Input: :math:`(N, in\_objects, in\_features)`
+        - Output: :math:`(N, out\_features)`
+
+    Examples::
+
+        >>> g = nn.Linear(60, 20)  # 60 is 2 * 30 (double the feature size per object)
+        >>> f = nn.Linear(20, 10)
+        >>> m = nn.Relation(f, g)
+        >>> input = autograd.Variable(torch.randn(32, 5, 30))
+        >>> output = m(input)
+    """
+    super(Relation, self).__init__()
+    self.f = f
+    self.g = g
+
+  def forward(self, x):
+    # Batch size, number of objects, feature size
+    b, o, c = x.size()
+    # Construct pairwise indices
+    i = Variable(torch.arange(0, o).long().repeat(o))
+    j = Variable(torch.arange(0, o).long().repeat(o, 1).t().contiguous().view(-1))
+    # Create pairwise matrix
+    x = torch.cat((torch.index_select(x, 1, i), torch.index_select(x, 1, j)), 2)
+    # Pack into batches and pass through g
+    x = self.g(x.view(b * o ** 2, 2 * c))
```

Comment:
@soumith : For higher resolution inputs (either spatial dimensions or channels), things go OOM quickly. One option is to specify a max_batch_size for the class (independent of the training process). The RN class can then accumulate the outputs of the `g` module by only sending max_batch_size number of pairs through `g`.